### PR TITLE
Encryption api

### PR DESCRIPTION
**Pull request by sueloverso**
Not ready for merging.  Just creating for convenience for @ddanderson and I.

### CODE REVIEW

**Comment by sueloverso** [comment]
autobuild: retest this please

**Comment by sueloverso** [comment]
autobuild: retest this please

**Comment by sueloverso** [comment]
@michaelcahill This branch is now ready for a first review.  @ddanderson and I are nearing the end.

**Comment by michaelcahill** [comment]
@sueloverso, can you please merge develop and re-run `dist/s_all` so that this branch can run through testing?

@keithbostic, I didn't get to this review today: can you please do a first pass?

File: dist/stat_data.py [hunk 1]
```diff
@@ -233,7 +233,6 @@ def __init__(self, name, desc, flags=''):
     LogStat('log_prealloc_max',
         'number of pre-allocated log files to create', 'no_clear,no_scale'),
     LogStat('log_prealloc_used', 'pre-allocated log files used'),
-    LogStat('log_reads', 'log read operations'),
     LogStat('log_release_write_lsn', 'log release advances write LSN'),
     LogStat('log_scan_records', 'records processed by log scan'),
     LogStat('log_scan_rereads', 'log scan records requiring two reads'),
```
Comment (keithbostic) [inline dist/stat_data.py]: @sueloverso, @michaelcahill: I can't imagine anyone is using the `log_reads` statistic, but we have a request from Amazon to never change existing statistics. Michael?

If we're really going to agree to never change statistics, we should probably put some process in place to guarantee that is the case from release-to-release.

**Comment by michaelcahill** [comment]
With statistics changes, I think we need to treat them like any other API changes: consider compatibility, call out changes in the upgrading section of the documentation, etc.  In this case, as noted, there should be no real problem here because we don't think anyone would be using this one.

**Comment by keithbostic** [comment]
@sueloverso, @michaelcahill, @ddanderson, I'm signed off, ready to merge when everyone else is.

**Comment by keithbostic** [comment]
@sueloverso, @ddanderson: one of the things that came out of a conversation Sue, @michaelcahill and I had last night is there isn't currently a need for the `WT_PAGE_ENCRYPTED` flag, that is, encryption isn't a characteristic of a page, it's a characteristic of the entire object, either every page is encrypted or no pages are encrypted, and `F_ISSET(WT_PAGE_ENCRYPTED)` tests could be replaced with `btree->kencryptor != NULL`.

Removing the `WT_PAGE_ENCRYPTED` flag from pages makes our backward/forward compatibility marginally better, as objects won't change.

That said, one of the things we've discussed is some future WiredTiger release will support an alter-table command that might turn on encryption for an existing object, leaving existing unencrypted pages alone. (Imagine a table to which sensitive data is being added.) So, the flag isn't necessary at the moment, but it might be necessary in the future.

So: my inclination is to leave things as they are, and I wanted to call out what I was thinking, please correct me if I'm wrong!

**Comment by michaelcahill** [comment]
@keithbostic, the usual pattern is to create the object, including setting the metadata, then try to open it (which does the validation).  If the open fails, the create is rolled back.

That is independent of encryption: if there is an issue with index creation, let's follow up on develop.

**Comment by michaelcahill** [comment]
Thanks, @sueloverso, @ddanderson and @keithbostic!  I'll merge...

**Pull request merged**